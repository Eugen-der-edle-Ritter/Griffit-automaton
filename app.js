window.onload = () => {
    /* GET CANVAS CONTEXT */
    const canvas = document.getElementById('canvas');
	const context = canvas.getContext('2d');
	
    /* PREPARING CANVAS AREA FOR FURTHER WORK */
	const width = canvas.width;
	const height = canvas.height;
	
	context.fillRect(0, 0, width, height);
	
	/* DEFINE STATE'S COLOR ARRAY */
	const states = ['#113', '#014', '#b16', '#54f', '#74f', '#fff', '#add', '#4fc', '#185', '#8c4', '#ef6', '#4bd', '#27f', '33f', '#059', '#567'];
	
	/* TRANSFORM HEX STRING INTO BYTE ARRAYS */
	const byteStates = colorTransform(states);
	
	/* DEFINE AND FILL 2d ARRAY OF CELL STATES BY RANDOM VALUES */
	let field = randomDraw(byteStates.length);

    /* TRANSFORMING CANVAS CELL'S STATES THROUGH THE TIME */
	setInterval(function () {
		let newField = [];
		
		for (let y = 0; y < height; y++) {
			let newRow = [];
			for (let x = 0; x < width; x++) {

                /* GET SUM OF NEIGHBOUR'S STATES */
				const stateSum =
                field[(y + height - 1) % height][x] +
                field[y][(x + width - 1) % width] +
                field[(y + height - 1) % height][(x + width - 1) % width] +
                field[(y + 1) % height][(x + 1) % width] +
                field[(y + height - 1) % height][(x + 1) % width] +
                field[(y + 1) % height][(x + width - 1) % width] +
                field[(y + 1) % height][x] +
                field[y][(x + 1) % width];

                /* CHANGE CELL'S STATE DEPENDING ON THE STATE'S SUM AND AUTOMATON RULES */
                if (field[y][x] === 0) {
					if (stateSum < 5) {
						newRow[x] = 0
					} else if (stateSum < 100) {
						newRow[x] = 2
					} else {
						newRow[x] = 3
					}
                } else if (field[y][x] === byteStates.length - 1) {
                    newRow[x] = 0;
                } else {
                    newRow[x] = Math.min(Math.floor(stateSum / 8) + 5, byteStates.length - 1);
                }
			}
			newField[y] = newRow;
		}
		field = newField;
		draw();
		
	}, 100);

    // CONVERTING HEX INTO BYTE ARRAY
	function colorTransform(colorArr) {
        const result = [];
        for (let i = 0; i < colorArr.length; i++) {
            result[i] = [
                parseInt(colorArr[i].charAt(1) + colorArr[i].charAt(1), 16),
                parseInt(colorArr[i].charAt(2) + colorArr[i].charAt(2), 16),
                parseInt(colorArr[i].charAt(3) + colorArr[i].charAt(3), 16)
            ];
        }
        return result;
    }
	
	/* RANDOM RENDERING CANVAS GRID */
    function randomDraw(stateCount) {
        const result = [];
        for (let y = 0; y < height; y++) {
            let row = [];
            for (let x = 0; x < width; x++) {
                row[x] = Math.floor(Math.random() * stateCount);
            }
            result[y] = row;
        }
        return result
    }
	
	/* RENDERING CANVAS GRID */
	function draw() {
		const imageData = context.getImageData(0, 0, width, height);
		for (let y = 0; y < height; y++) {
			for (let x = 0; x < width; x++) {
				const offset = (width * y + x) * 4;
				imageData.data[offset] = byteStates[field[y][x]][0];
				imageData.data[offset + 1] = byteStates[field[y][x]][1];
				imageData.data[offset + 2] = byteStates[field[y][x]][2];
			}
		}
		context.putImageData(imageData, 0, 0);	
	}
}