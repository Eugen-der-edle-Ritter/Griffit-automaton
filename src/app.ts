import '../public/style.css';
import { cell } from './modules/types';
import { randomFill, transitionFill } from './modules/filling';

/* INITILIZE AUTOMATA STATE */
const statesArr: string[] = [
  '#113',
  '#014',
  '#b16',
  '#54f',
  '#74f',
  '#fff',
  '#add',
  '#4fc',
  '#185',
  '#8c4',
  '#ef6',
  '#4bd',
  '#27f',
  '#33f',
  '#059',
  '#567',
];
const states: number = statesArr.length;

/* INTERFACE FOR GLOBAL CANVAS AUTOMATON CELL */
interface ICell {
  width: number;
  height: number;
}

window.onload = () => {
  /* GET CANVAS CONTEXT */
  const canvas: HTMLCanvasElement = document.getElementById(
    'canvas'
  ) as HTMLCanvasElement;
  const context: CanvasRenderingContext2D | null = canvas.getContext('2d');

  const { width, height }: ICell = canvas;

  /* DRAW EMPTY RECTANGLE */
  context!.fillRect(0, 0, width, height);

  /* SUBSCRIBE ON TRANSITION-TYPE TOGGLER */
  const ruleSelector: HTMLElement | null =
    document.querySelector('.ruleSelector');
  if (ruleSelector) {
    ruleSelector.addEventListener('change', (e): void => {
      const newRule: string = (e.target as HTMLTextAreaElement).value;
      clearInterval(timer);
      field = randomFill(byteArr.length, height, width);
      timer = drawAutomata(newRule);
    });
  }

  /* TRANSFORM HEX STRING INTO BYTE ARRAYS */
  const byteArr: number[][] = colorTransform(statesArr);

  console.log(byteArr, 'hi')

  /* DEFINE AND FILL 2d ARRAY OF CELL STATES BY RANDOM VALUES */
  let field: cell[][] = randomFill(byteArr.length, height, width);

  /* TRANSFORMING CANVAS CELL'S STATES THROUGH THE TIME */
  let timer = drawAutomata('hash');

  function drawAutomata(rule: string) {
    const timer = setInterval(() => {
      field = transitionFill(rule, field, states, height, width);
      draw();
    }, 50);
    return timer;
  }

  /* CONVERTING HEX INTO BYTE ARRAY */
  function colorTransform(colorArr: string[]): number[][] {
    return colorArr.map((el, i) => {
      return [
        parseInt(colorArr[i].charAt(1) + colorArr[i].charAt(1), 16),
        parseInt(colorArr[i].charAt(2) + colorArr[i].charAt(2), 16),
        parseInt(colorArr[i].charAt(3) + colorArr[i].charAt(3), 16),
      ];
    });
  }

  /* RENDERING CANVAS GRID */
  function draw(): void {
    const imageData = context!.getImageData(0, 0, width, height);
    for (let y = 0; y < height; y++) {
      for (let x = 0; x < width; x++) {
        const offset = (width * y + x) * 4;
        imageData.data[offset] = byteArr[field[y][x]][0];
        imageData.data[offset + 1] = byteArr[field[y][x]][1];
        imageData.data[offset + 2] = byteArr[field[y][x]][2];
      }
    }
    context!.putImageData(imageData, 0, 0);
  }
};
