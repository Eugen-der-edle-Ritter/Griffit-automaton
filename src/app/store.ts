import { configureStore, ThunkAction, AnyAction } from '@reduxjs/toolkit';

import cellsReducer, { setCells } from '@/components/Main/Home/cellsSlice';
import playReducer from '@/components/Main/Home/ControlPanel/playToggler/playSlice';
import ruleReducer from '@/components/Main/Home/ControlPanel/ruleSwitcher/ruleSlice';
import speedReducer from '@/components/Main/Home/Accelerator/speedSlice';
import colorReducer from '@/components/Main/Home/ControlPanel/colorRandomizer/colorSlice';

import { width, height, statesCount } from '@/constants/constants';

import { randomFill, transitionFill } from '@/modules/filling';

export const store = configureStore({
  reducer: {
    cells: cellsReducer,
    play: playReducer,
    rule: ruleReducer,
    speed: speedReducer,
    colors: colorReducer,
  },
  middleware: (getDefaultMiddleware) => [
    ...getDefaultMiddleware({
      serializableCheck: false,
    }),
  ],
});

// store.subscribe(() => {
//   localStorage.setItem(
//     'cellsState',
//     JSON.stringify(store.getState().cells.present.value)
//   );
// });

export type RootState = ReturnType<typeof store.getState>;

export const randomUpdate =
  (): ThunkAction<void, RootState, unknown, AnyAction> => (dispatch) => {
    dispatch(setCells(randomFill(statesCount, height, width)));
  };

export const recursiveUpdate =
  (): ThunkAction<void, RootState, unknown, AnyAction> =>
  (dispatch, getState) => {
    const { play, speed, rule, cells } = getState();
    if (play.value) {
      setTimeout(() => {
        dispatch(
          setCells(
            transitionFill(
              rule.value,
              cells.present.value,
              statesCount,
              height,
              width
            )
          )
        );
        dispatch(recursiveUpdate());
      }, Math.floor(70 / speed.value) + 1);
    }
  };

export const runTimer =
  (): ThunkAction<void, RootState, unknown, AnyAction> =>
  (dispatch, getState) => {
    const isPlaying = getState().play.value;
    if (isPlaying) {
      dispatch(recursiveUpdate());
    }
  };
