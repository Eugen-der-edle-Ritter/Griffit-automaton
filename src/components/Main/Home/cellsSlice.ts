import { createSlice } from '@reduxjs/toolkit';
import undoable from 'redux-undo';

import { Cell } from '@/modules/types';

export const cellsSlice = createSlice({
  name: 'cells',
  initialState: {
    value: [] as Cell[][],
  },
  reducers: {
    setCells: (state, action) => {
      state.value = action.payload;
    },
  },
});

// Action creators are generated for each case reducer function
export const { setCells } = cellsSlice.actions;

export default undoable(cellsSlice.reducer);
