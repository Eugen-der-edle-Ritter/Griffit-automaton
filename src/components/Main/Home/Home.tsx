import React, { FC } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import styled from '@emotion/styled';

import { Cell, Rule } from '@/types';
import { randomUpdate, RootState } from '@/app/store';

import { width, height, scale } from '@/constants/constants';

import { Accelerator } from './Accelerator/Accelerator';
import { ControlPanel } from './ControlPanel/ControlPanel';
import { Display } from './Display/Display';
// import { DisplayBlocks as Display } from './Display/DisplayBlocks';

export const Home: FC = () => {
  const [cellsState, rule]: [Cell[][], Rule] = useSelector(
    (state: RootState) => [state.cells.present.value, state.rule.value]
  );
  const dispatch = useDispatch();

  React.useEffect(() => {
    dispatch(randomUpdate());
  }, [rule]);

  return (
    <>
      <Header>
        <Title>Cyclic cellular automaton</Title>
      </Header>
      <Main>
        <ControlPanel />
        <Display cellsState={cellsState} width={width} height={height} />
        <Accelerator />
      </Main>
    </>
  );
};

const Header = styled.header``;
const Main = styled.main`
  max-width: ${width * scale}px;
  margin: 0 auto;
  display: flex;
  flex-direction: column;
  align-items: center;
`;
const Title = styled.h1`
  margin: 3rem;
  text-align: center;
  font-size: 2.5rem;
`;
