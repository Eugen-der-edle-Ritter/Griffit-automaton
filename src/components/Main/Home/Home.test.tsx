import React from 'react';
import { store } from '@/app/store';
import { Provider, useDispatch, useSelector } from 'react-redux';

import { render } from '@testing-library/react';

import { Home } from './Home';

jest.mock('react-redux', () => ({
  __esModule: true,
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

/* mocking useEffect */
const useEffect = jest.spyOn(React, 'useEffect');

const mockUseEffect = () => {
  useEffect.mockImplementationOnce((f) => f());
};

it('Dispatch action inside useEffect hook', () => {
  const mockDispatch = jest.fn();

  (useDispatch as jest.Mock).mockReturnValue(mockDispatch);

  mockUseEffect();
  expect(mockDispatch).toHaveBeenCalled();
});

it('Home section renders correctly', () => {
  const mockDispatch = jest.fn();

  (useDispatch as jest.Mock).mockReturnValue(mockDispatch);
  (useSelector as jest.Mock).mockReturnValue([[], 'Hash']);
  const { asFragment } = render(
    <Provider store={store}>
      <Home />
    </Provider>
  );

  expect(asFragment).toMatchSnapshot();
});
