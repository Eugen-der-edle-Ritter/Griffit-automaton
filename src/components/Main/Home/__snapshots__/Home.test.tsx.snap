// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home section renders correctly 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Home />
</Provider>
`;
